File: interface.rb
#!/usr/bin/env ruby

require 'ffi'

module LibObject
  extend FFI::Library

  ffi_lib './libObject.so'
  attach_function :object_int_init, [:int], :pointer
  attach_function :object_int_destroy, [:pointer], :void
  attach_function :object_int_n_get, [:pointer], :int
  attach_function :object_int_n_put, [:pointer, :int], :pointer
  attach_function :object_int_id, [:pointer], :string

  attach_function :object_float_init, [:float], :pointer
  attach_function :object_float_destroy, [:pointer], :void
  attach_function :object_float_n_get, [:pointer], :float
  attach_function :object_float_n_put, [:pointer, :float], :pointer
  attach_function :object_float_id, [:pointer], :string
end

class ObjectInt
  include LibObject
  include ObjectSpace # for destructor
  def initialize(n = 0); @ptr = object_int_init(n); end
  def destroy; object_int_destroy(@ptr); end
  def n; object_int_n_get(@ptr); end
  def n=(m); object_int_n_put(@ptr, m); end
  def id; object_int_id(@ptr); end

  def ObjectInt.finalize(id)
    obj = ObjectSpace._id2ref(id)
    obj.destroy
  end
end

class ObjectFloat
  include LibObject
  include ObjectSpace # for destructor
  def initialize(n = 0); @ptr = object_float_init(n); end
  def destroy; object_float_destroy(@ptr); end
  def n; object_float_n_get(@ptr); end
  def n=(m); object_float_n_put(@ptr, m); end
  def id; object_float_id(@ptr); end

  def ObjectFloat.finalize(id)
    obj = ObjectSpace._id2ref(id)
    obj.destroy
  end
end

if __FILE__ == $0
  
  puts "Creating three new objects..."
  obj_i1 = ObjectInt.new 123
  obj_i2 = ObjectInt.new 987
  obj_f1 = ObjectFloat.new 3.14

  puts "Integer object 1 is #{obj_i1.id} with n = #{obj_i1.n}"
  puts "Integer object 2 is #{obj_i2.id} with n = #{obj_i2.n}"
  puts "Float object 1 is #{obj_f1.id} with n = #{obj_f1.n}"

  puts "Updating n for objects"
  obj_i1.n = 321
  obj_i2.n = 789
  obj_f1.n = 4.13
  
  puts "Integer object 1 is #{obj_i1.id} with n = #{obj_i1.n}"
  puts "Integer object 2 is #{obj_i2.id} with n = #{obj_i2.n}"
  puts "Float object 1 is #{obj_f1.id} with n = #{obj_f1.n}"  

end

File: interface.js
#!/usr/bin/env node

var ffi = require('ffi');
var ref = require('ref');

var voidPtr = ref.refType('void');

var libObject = ffi.Library('./libObject.so', {
  "object_int_init": [voidPtr , ['int']],
  "object_int_destroy": ['void' , [voidPtr]],
  "object_int_n_get": ['int' , [voidPtr]],
  "object_int_n_put": [voidPtr , [voidPtr, 'int']],
  "object_int_id": ['string', [voidPtr]],

  "object_float_init": [voidPtr , ['float']],
  "object_float_destroy": ['void' , [voidPtr]],
  "object_float_n_get": ['float' , [voidPtr]],
  "object_float_n_put": [voidPtr , [voidPtr, 'float']],
  "object_float_id": ['string' , [voidPtr]],
});

var ObjectInt = function(n) { 
    this.ptr = libObject.object_int_init(n);

    this.delete = function() {
      if (this.ptr) libObject.object_int_id(this.ptr);
      this.ptr = undefined;
    };
};

ObjectInt.prototype = {
  get n () { 
    if (this.ptr)
      return libObject.object_int_n_get(this.ptr);  
    return undefined;
  },
  set n(m) { 
    if (this.ptr) {
      libObject.object_int_n_put(this.ptr, m); 
      return this; 
    }
    return undefined;
  },
  get id () {
    if (this.ptr)
      return libObject.object_int_id(this.ptr);
    return undefined;
  }
}

var ObjectFloat = function (n) {
  this.ptr = libObject.object_float_init(n);

  this.delete = function () {
    if (this.ptr) libObject.object_float_id(this.ptr);
    this.ptr = undefined;
  };
};

ObjectFloat.prototype = {
  get n() {
    if (this.ptr)
      return libObject.object_float_n_get(this.ptr);
    return undefined;
  },
  set n(m) {
    if (this.ptr) {
      libObject.object_float_n_put(this.ptr, m);
      return this;
    }
    return undefined;
  },
  get id() {
    if (this.ptr)
      return libObject.object_float_id(this.ptr);
    return undefined;
  }
}


var obj_i1 = new ObjectInt(123);
var obj_i2 = new ObjectInt(987);
var obj_f1 = new ObjectFloat(3.14);

console.log("Integer object 1 is " + obj_i1.id + " with n = " + obj_i1.n);
console.log("Integer object 2 is " + obj_i2.id + " with n = " + obj_i2.n);
console.log("Float object 1 is " + obj_f1.id + " with n = " + obj_f1.n);

console.log("Updating n for objects");
obj_i1.n = 321;
obj_i2.n = 789;
obj_f1.n = 4.13;

console.log("Integer object 1 is " + obj_i1.id + " with n = " + obj_i1.n);
console.log("Integer object 2 is " + obj_i2.id + " with n = " + obj_i2.n);
console.log("Float object 1 is " + obj_f1.id + " with n = " + obj_f1.n);

obj_i1.delete();
obj_i2.delete();
obj_f1.delete();